File: 0020-valid-parentheses/solution.go
var (
    mapPairs = map[rune]rune{
        ')': '(',
        '}': '{',
        ']': '[',
    }
)

func isValid(s string) bool {
    stack := []rune{}
    for _, v := range s {
        switch v {
            case ')', '}', ']':
                if len(stack) - 1 < 0 {
                    return false
                }

                popped := stack[len(stack) - 1]
                if mapPairs[v] != popped {
                    return false
                }
                stack = stack[:len(stack) - 1]
                continue
        }

        stack = append(stack, v)
    }

    return len(stack) == 0
}

File: 0020-valid-parentheses/solution.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let stack = [];
    for (let i = 0; i < s.length; i++) {
        if (s[i] === ')') {
            const last = stack.pop();
            if (last !== '(') {
                return false;
            }
        } else if (s[i] === '}') {
            const last = stack.pop();
            if (last !== '{') {
                return false;
            }
        } else if (s[i] === ']') {
            const last = stack.pop();
            if (last !== '[') {
                return false;
            }
        } else {
            // opening bracket
            stack.push(s[i]);
        }
    }

    return stack.length === 0;
};
